cmake/SeracMacros.cmake

```diff
@@ -218,3 +218,176 @@ macro(serac_configure_file _source _target)
+        string(REPLACE "${arg_PREFIX}" "" _stripped_path "${arg_INPUT}")
+        set(${arg_OUTPUT_VAR} "${_stripped_path}")
+    else()
+        set(${arg_OUTPUT_VAR} "${arg_INPUT}")
+    endif()
+endmacro()
+
+##------------------------------------------------------------------------------
+## serac_remove_string_prefix
+##
+## This macro fills OUTPUT_VAR with the subdirectory relative to 
+## "<lower case project name>/src"
+##
+## OUTPUT_VAR - Possibly altered output string
+##
+##------------------------------------------------------------------------------
+macro(serac_get_src_subdirectory)
+
+    set(options)
+    set(singleValueArgs OUTPUT_VAR)
+    set(multiValueArgs)
+
+    # Parse the arguments to the macro
+    cmake_parse_arguments(arg
+         "${options}" "${singleValueArgs}" "${multiValueArgs}" ${ARGN})
+
+    if(NOT DEFINED arg_OUTPUT_VAR)
+         message(FATAL_ERROR "OUTPUT_VAR is required")
+    endif()
+
+    string(TOLOWER ${PROJECT_NAME} _project_name)
+
+    # Convert both paths to absolute
+    file(REAL_PATH "${PROJECT_SOURCE_DIR}/src/${_project_name}" abs_base_dir)
+    file(REAL_PATH "${CMAKE_CURRENT_SOURCE_DIR}" abs_full_path)
+
+    # Check if full path starts with the base directory
+    string(FIND "${abs_full_path}" "${abs_base_dir}" found_pos)
+
+    if(NOT found_pos EQUAL 0)
+        message(FATAL_ERROR "The full path '${abs_full_path}' does not start with the base directory '${abs_base_dir}'")
+    endif()
+
+    # Strip base directory from full path
+    serac_remove_string_prefix(PREFIX "${abs_base_dir}/"
+                               INPUT "${abs_full_path}"
+                               OUTPUT_VAR ${arg_OUTPUT_VAR})
+endmacro(serac_get_src_subdirectory)
+
+
+##------------------------------------------------------------------------------
+## serac_write_unified_header
+##
+## This macro writes the unified header (<lowered PROJECT_NAME/<lowered NAME>.hpp)
+## to the build directory for the given NAME with the given HEADERS included
+## inside of it.
+##
+## NAME - The name of the unified header.
+## HEADERS - Headers to be included in the header.
+## NO_PATH_MODIFICATION - ON/OFF(default) Stops include path modification if on,
+##    used for the project-level unified header
+##
+##------------------------------------------------------------------------------
+# List to hold all unified headers to be later used to create a master unified header
+set(_serac_unified_headers "" CACHE STRING "")
+macro(serac_write_unified_header)
+
+    set(options)
+    set(singleValueArgs NAME NO_PATH_MODIFICATION)
+    set(multiValueArgs HEADERS EXCLUDE)
+
+    # Parse the arguments to the macro
+    cmake_parse_arguments(arg
+         "${options}" "${singleValueArgs}" "${multiValueArgs}" ${ARGN})
+
+    if(NOT DEFINED arg_NO_PATH_MODIFICATION)
+        set(arg_NO_PATH_MODIFICATION OFF)
+    endif()
+
+    string(TOLOWER ${arg_NAME} _lcname)
+    string(TOLOWER ${PROJECT_NAME} _project_name)
```

Comment:
might be worth making these variable names consistent. they're both lower case, so maybe they could both have `_lc`. or personally i think it makes more sense to rename`_lcname` to `_header_name`